### ISSUE TITLE
Difficulty in reproducing results from paper

### ISSUE THREAD

**Issue by nikriz1** [jeffacce/dynamo_ssl#6]
Hello,

Thank you for sharing this code and your research. 
I’ve been experimenting to test dynamo in the Franka Kitchen environment but encountered difficulties in reproducing the policy performances reported in the paper.

I’ve tried with the encoder trained using the code and the provided checkpoint, then trained VQ-BeT on the encoders.
All configurations except paths are kept same as the code provided, which is identical to parameters from papers. 

added below code in `models.encoder.multiview_stack.MultiviewStack`
to load provided checkpoint to encoder.
```
from collections import OrderedDict
from .resnet import resnet18

...

        if type(encoders[0]) == OrderedDict: 
            resnet = resnet18(
                    pretrained=False,
                    output_dim=output_dim,
                    unit_norm=False,
                )
            resnet.load_state_dict(encoders[0])
            encoders = [resnet]
```

Below are wandb logs of SSL
![ssl](https://github.com/user-attachments/assets/6e28015d-94ed-4d64-9f6f-c3663d413b8f)
*sorry for unrecognizable values and some missing plots

wandb logs of BeT
orange is trained encoder, pink is provided checkpoint.
![bet](https://github.com/user-attachments/assets/eaa18444-15ca-4ee7-80bd-ba5a149f0105)
*eval total_loss mins at 9.6 then overfits to 11+

To my understanding, eval_on_env should match the Table 1's dynamo results from the paper.
Any idea what could have been gone wrong about my experiments?

The experiments are done with
Ubuntu 20.04, python 3.8, torch 2.4.0+cu121, CUDA 12.2, single H100
with all packages installed by provided conda_env.yml except CUDA related ones.

I would greatly appreciate any guidance you could offer to help me resolve this issue.

**Comment by nikriz1** [comment]
Above experiments were finished with 
checkpoint max 1.95 final 1.15 
trained max 2.5 final 1.2

Other tasks results were
blockpush max moved mean 1.55
blockpush max entered mean 0.52 (paper 0.65)
pushT max max coverage mean 0.62943 
pushT max final coverage mean 0.57945 (paper 0.66)
libero max eval on env 0.85 (paper 0.93)

all used re-trained encoders, shows overfitting eval total loss.

**Comment by jeffacce** [comment]
Hey @nikriz1, thanks for checking out our work and running the experiments!

# Sim kitchen
Yeah sim kitchen shouldn't be this low. This sounds like the policy is out of distribution perhaps?
- Is it rendering with `glfw` or `egl`? There might be differences between CPU/GPU rendering.
- Could you give us the MuJoCo version?
- Which tasks are completed?

We've just run some repro runs with freshly trained encoders and the uploaded checkpoint:
<img width="553" alt="sim kitchen" src="https://github.com/user-attachments/assets/4d3a71f8-00b8-4577-83ab-9ddcb4bbd0b8">

The highlighted run (pink) is with a randomly initialized, frozen ResNet18 (random features), and the other runs are with either the provided checkpoint or a freshly trained encoder. Sim kitchen is a relatively open-loop environment, and in our experiments we've seen that just random or ImageNet pretrained features should do above 3. We couldn't reproduce your 1.2 result on our machine. We're using MuJoCo 2.1.0 with `MUJOCO_GL=egl` and the provided conda env.

The 3.64 run reported in the paper is included below.
<img width="2070" alt="sim kitchen" src="https://github.com/user-attachments/assets/341850db-33cb-4437-a816-ef40e8f0ce66">

# Blockpush, Push-T, LIBERO
Yep these results sound reasonable. PyTorch is deterministic up to the same hardware, and different machines will yield slightly different results. We fix the seed and use deterministic CUDA algorithms, so the same hardware will give you deterministic encoder training, and very close to deterministic policy eval results, but results across machines won't be exactly the same. (See: https://pytorch.org/docs/stable/notes/randomness.html)

We did run a few repros on different cards and found the following:

## Blockpush
Blocks entered: 0.65 - 0.79
<img width="541" alt="blockpush" src="https://github.com/user-attachments/assets/addea0a3-bd5c-4504-9187-16d944ed17e8">

Run reported in paper: 0.65
<img width="2054" alt="blockpush" src="https://github.com/user-attachments/assets/ff245b05-2596-4181-8a32-902813f48248">

## Push-T
Final coverage 0.59 - 0.66
<img width="540" alt="pusht" src="https://github.com/user-attachments/assets/8e227db0-3377-404f-ac35-d4f458412755">

Run reported in paper: 0.66
<img width="2071" alt="pusht" src="https://github.com/user-attachments/assets/92ed43f7-b330-4f63-8750-691231515fde">

## LIBERO Goal
Result 0.78 - 0.93
<img width="539" alt="libero" src="https://github.com/user-attachments/assets/1f65a35d-dd08-4618-9b73-a4317a852748">

Run reported in paper: 0.93
<img width="2066" alt="libero" src="https://github.com/user-attachments/assets/f18506ab-e185-4c8f-a430-67887fd32a87">

Let me know if you have any questions!

**Comment by nikriz1** [comment]
Thank you so much for your kind and detailed response!

We also have MuJoCo 2.1.0 binary installed,
along with python environment in `conda_env.yml` (`gym==0.23.1, mujoco==2.3.7, mujoco_py==2.1.2.14`)
and rendering setting specified in README (`MUJOCO_GL=egl`) 

My re-trained encoder checkpoint and completed tasks (max 1.2) for kitchen environment are here:
- [encoder.pt](https://drive.google.com/file/d/11nx3KceBMxrfCbTsDBJJnoJh77gACOYJ/view?usp=sharing)  
- [completion_idx_final.json](https://github.com/user-attachments/files/17446699/completion_idx_final.json)

I'll also try to run the experiments on a different machine as well. 
Thank you!

**Comment by nikriz1** [comment]
Kitchen worked fine (3.75) on different machine!
